aten/src/ATen/quantized/Quantizer.cpp

```diff
@@ -0,0 +1,100 @@
+  AT_ASSERTM(!self.is_variable(), "_internal_get_QTensorImpl: should not be a variable");
+  // TODO: uncomment after is_quantized() is implmeented
+  // AT_ASSERTM(self.is_quantized(), "_internal_get_QTensorImpl: not a quantized tensor");
+  return static_cast<QTensorImpl*>(self.unsafeGetTensorImpl());
+}
+
+inline QTensor new_qtensor(
+    IntArrayRef sizes, const TensorOptions& options, bool is_variable, QuantizerPtr quantizer) {
+  AT_ASSERT(options.device().is_cpu());
+
+  native::check_size_nonnegative(sizes);
+  auto* allocator = at::getCPUAllocator();
+  int64_t nelements = at::prod_intlist(sizes);
+  auto dtype = options.dtype();
+  AT_ASSERT(isQIntType(typeMetaToScalarType(dtype)));
+  auto storage_impl = c10::make_intrusive<StorageImpl>(
+    dtype,
+    nelements,
+    allocator->allocate(nelements * dtype.itemsize()),
+    allocator,
+    /*resizable=*/true);
+  auto tensor = detail::make_tensor<QTensorImpl>(
+      storage_impl, at::CPUTensorId(), is_variable, quantizer);
+  get_qtensorimpl(tensor)->set_sizes_contiguous(sizes);
+  return tensor;
+}
+
+qint8 quantize_uint8(float scale, uint8_t zero_point, float value) {
+  const int32_t qmin = std::numeric_limits<uint8_t>::min();
+  const int32_t qmax = std::numeric_limits<uint8_t>::max();
+
+  // std::nearbyint results in nearest integer value according to the current rounding
+  // mode and the default rounding mode is rounds to even in half-way cases in most
+  // popular processor architectures like x86 and ARM.
+  // This is typically faster than an alternatives like std::round that rounds half-way
+  // cases away from zero, and can be consistent with SIMD implementations for example
+  // in x86 using _mm512_cvtps_epi32 or mm512_round_ps with _MM_FROUND_CUR_DIRECTION
+  // option that also follow the current rounding mode.
+  auto r = zero_point + static_cast<int32_t>(std::nearbyint(value / scale));
+  r = std::max(r, qmin);
+  r = std::min(r, qmax);
+  return static_cast<qint8>(r);
+}
+
+QTensor PerTensorAffineQuantizer::quantize(RealTensor tensor) {
+  IntArrayRef sizes = tensor.sizes();
+  // Here we need a std::intrusive_ptr<Quantizer>.. but actually "this" is the quantizer that
+  // can be reused, so I'm using intrusive_from_this here
+  QTensor qv = new_qtensor(sizes, tensor.options().dtype(at::kQInt8), tensor.is_variable(), intrusive_from_this());
+  auto qvd = qv.data<qint8>();
+  tensor.contiguous();
+  const float* svd = tensor.data<float>();
+  for (int i = 0; i < tensor.numel(); ++i) {
+    qvd[i] = quantize_uint8(scale_, zero_point_, svd[i]);
+  }
+  return qv;
+}
+
+RealTensor PerTensorAffineQuantizer::dequantize(QTensor tensor) {
+  std::vector<int64_t> sizes = tensor.sizes().vec();
+  at::TensorOptions real_options = tensor.options().dtype(at::kFloat);
+
+  RealTensor rv = at::empty(sizes, real_options);
+  tensor.contiguous();
+  const auto* qvd = tensor.data<qint8>();
+  float* rvd = rv.data<float>();
```

Comment:
this will work only CPU device, so I'd assert that device in tensor options is CPU (and we'd need to add dispatch for other devices in the future